external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+}
+
+void replaceBF16InRegion(Region &region, LLVMTypeConverter &converter) {
+  for (Block &block : region.getBlocks())
+    replaceBF16InBlock(block, converter);
+}
+
+void replaceBF16InOp(Operation *op, LLVMTypeConverter &converter) {
+  if (auto func = dyn_cast<LLVM::LLVMFuncOp>(op)) {
+    auto funcType = func.getFunctionType();
+    func.setType(converter.convertType(funcType));
+    for (auto arg : func.getArguments())
+      arg.setType(converter.convertType(arg.getType()));
+  } else {
+    for (unsigned idx = 0; idx < op->getNumOperands(); idx++) {
+      auto type = converter.convertType(op->getOperand(idx).getType());
+      op->getOperand(idx).setType(type);
+    }
+    for (unsigned idx = 0; idx < op->getNumResults(); idx++) {
+      auto type = converter.convertType(op->getResult(idx).getType());
+      op->getResult(0).setType(type);
+    }
+  }
+  for (Region &region : op->getRegions())
+    replaceBF16InRegion(region, converter);
+  return;
+}
+
+} // namespace LLVM
+} // namespace mlir
+
+void mlir::LLVM::populateBF16ToLLVMConversionPatterns(
+    LLVMTypeConverter &converter, RewritePatternSet &patterns) {
+  MLIRContext *ctx = converter.getDialect()->getContext();
+  // AMD GPUs don't have a backend that understands bfloat, even though LLVM's
+  // frontend does. Remove this if/when that changes. Note that adding
+  // conversions after the default constructor runs gives them priority
+  // over the defaults.
+  Type llvmI16 = converter.convertType(IntegerType::get(ctx, 16));
+  Type bf16 = mlir::BFloat16Type::get(ctx);
+  converter.addConversion(
+      [llvmI16](mlir::BFloat16Type type) -> Type { return llvmI16; });
+
+  // Override for vector/struct types since they get caught by
+  // isCompatibleType(), which doesn't convert the element type
+  converter.addConversion(
+      [llvmI16, bf16](mlir::VectorType type) -> Optional<Type> {
+        if (type.getElementType() == bf16 && type.getRank() == 1)
+          return type.clone(llvmI16);
+        return llvm::None; // continue search
+      });
+
+  converter.addConversion([&](LLVM::LLVMStructType type) -> Optional<Type> {
+    bool converted = false;
+    SmallVector<Type> convertedElemTypes;
+    convertedElemTypes.reserve(type.getBody().size());
+    for (auto t : type.getBody()) {
+      SmallVector<Type, 1> element;
+      if (failed(converter.convertType(t, element)))
+        return llvm::None;
+      convertedElemTypes.push_back(element[0]);
+      if (t != element[0])
+        converted = true;
+    }
+
+    if (!converted)
+      return type;
+
+    if (type.isIdentified()) {
+      auto convertedType = LLVM::LLVMStructType::getIdentified(
+          type.getContext(), ("_Converted_" + type.getName()).str());
+      unsigned counter = 1;
+      while (convertedType.isInitialized()) {
+        convertedType = LLVM::LLVMStructType::getIdentified(
+            type.getContext(),
+            ("_Converted_" + std::to_string(counter++) + type.getName()).str());
```

Comment:
I think there might be some usage of `Twine` possible here to avoid the `to_string`